lib/ex_doc/autolink.ex

```diff
@@ -389,71 +394,105 @@ defmodule ExDoc.Autolink do
 
   ## Internals
 
-  defp module_url(module, config) do
-    if module == config.current_module do
-      "#content"
-    else
-      if Refs.public?({:module, module}) do
-        module_url(tool(module), module, config)
-      end
+  defp module_url(module, module_string, mode, config, mix_task? \\ false)
+
+  defp module_url(module, _module_string, _mode, %{current_module: module}, _mix_task?),
+    do: "#content"
+
+  defp module_url(module, module_string, mode, config, mix_task?) do
+    case {mode, Refs.get_visibility({:module, module})} do
+      {_, :public} ->
+        module_url_app(tool(module), module, config)
+
+      {:regular_link, :undefined} ->
+        nil
+
+      {_, visibility} ->
+        message =
+          if mix_task? do
+            "documentation references \"mix #{module_string}\" but such task is #{visibility}"
+          else
+            "documentation references module \"#{module_string}\" but it is #{visibility}"
```

Comment:
I'm starting to think that _kind_ in "documentation references kind (...)" isn't that useful after all and we could show the prefix (if any) instead:

    documentation references module "Unknown" but it is undefined
    documentation references function "Unknown.foo/1" but it is undefined
    documentation references type "Unknown.t/1" but it is undefined
    documentation references callback "Unknown.handle_foo/1" but it is undefined

we could do:

    documentation references "Unknown" but it is undefined
    documentation references "Unknown.foo/1" but it is undefined
    documentation references "t:Unknown.t/1" but it is undefined
    documentation references "c:Unknown.handle_foo/1" but it is undefined

What's nice is we avoid the function vs macro problem (not that it is a big deal). I think this could potentially simplify some code too, what do you all think? If that sounds reasonable, I'm happy to tackle it later.

I guess the downside, and I think this was the reason it is like that in the first place, is e.g. `t:` is specific to ExDoc, so if we'd eventually support EDoc docs with ExDoc, using the `t:` in the warning would be a bit awkward. But on the flip side, the warning comes from ExDoc so it's reasonable to use conventions that make sense for _it_.

If we do this, I think we'd also simplify the code around Mix tasks, I think what we have on this patch:

    documentation references \"mix ...\" but such task is ..."

is fine, but I'd personally be totally fine with:

    documentation references \"mix ...\" but it is ..."

I don't think we lose much because we have "mix " in the warning already. See the other comment about mix tasks too.

Reply:
I finished before you reviewed this code another PR that refactor the warn and maybe_warn functions. What you suggest could easily be done there. I will submit the PR and we can have this conversation over there

Reply:
Could you please copy and past this comment over there? https://github.com/elixir-lang/ex_doc/pull/1235
So I can make this issue as resolved. This PR will be based on that one if merged